selene_sdk/samplers/random_positions_sampler.py

```diff
@@ -151,86 +159,214 @@ def __init__(self,
-            self._sample_from_mode["test"] = SampleIndices(
-                test_indices, test_weights)
-
-            tr_indices, tr_weights = get_indices_and_probabilities(
-                self.interval_lengths, select_indices[test_indices_end:])
-            self._sample_from_mode["train"] = SampleIndices(
-                tr_indices, tr_weights)
+            test_prop_N = self._genome_n_bases * self._test_holdout
+            validation_prop_N = self._genome_n_bases * self._validation_holdout
+            training_prop_N = \
+                self._genome_n_bases * (1 - self._test_holdout - self._validation_holdout)
+            return int(test_prop_N), int(validation_prop_N), int(training_prop_N)
         else:
-            tr_indices, tr_weights = get_indices_and_probabilities(
-                self.interval_lengths, select_indices[n_indices_validate:])
-            self._sample_from_mode["train"] = SampleIndices(
-                tr_indices, tr_weights)
+            validation_prop_N = self._genome_n_bases * self._validation_holdout
+            training_prop_N = self._genome_n_bases * (1 - self._validation_holdout)
+            return int(validation_prop_N), int(training_prop_N)
+
+    # Should be updated to be dynamic for the genome that is being sampled.
+    # If the genome size is not a multiple of 250 million, then we do it
+    # differently ..
+    def _psuedoshuffle(self):
+        total = self._genome_n_bases
+        shuffled_sample = np.arange(total, dtype=np.int64)
+        jump = 250000000
+        num_steps = int(total / jump)
+        start = 0
+        for i in range(num_steps):
+            np.random.shuffle(shuffled_sample[start : start + jump])
+            start = start + jump
+
+        start = int(jump / 2)
+        for i in range(num_steps - 1):
+            np.random.shuffle(shuffled_sample[start : start + jump])
+            start = start + jump
+
+        return shuffled_sample
+
+    # Samples for each mode are predetermined based on the sample size for
+    # each mode. This imposes an up front time constraint, dependent on each
+    # sample size and the size of the entire genome.  # Subsequently, sampling is quick.
+
+    # There is a space constraint in the first phase of picking the samples
+    # However, this constraint dwindles after samples are set and even further as
+    #  samples are taken and subsequently removed.
+
+    # If the genome is above 500 million base pairs, only a psuedoshuffle is
+    # performed on the genome to perform partitions. This is due to time
+    # constraints on shuffling larger sets.
+
+    # Samples are taken without replacement. However, if the sample size is
+    # larger than the proportion specified, there will be duplicate sample
+    # items. If this is not the intended behavior, the sample size should be
+    # chosen so as to not be larger than the proportion of the genome.
+
+    def _partition_by_proportion(self):
```

Comment:
this is a super long function, can we refactor or modularize duplicate code for each mode or something to shorten it?

Reply:
The repeated code updates integers (`N_validation`, `start`), which I do not think I can update in a function directly (since integers are passed by copy in Python I believe?) 

However, maybe the lines:
get_N = min(validation_prop_N, N_validation)
self._partition_ixs["validate"] = genome_positions_arr[start : start + get_N]

can we be modularized.. Thoughts? I had the same concern but I was not sure if a two line function would be helpful here